drivers/ukbus/platform/platform_bus.c

```diff
@@ -261,22 +261,33 @@ __vaddr_t uk_bus_pf_devmap(__paddr_t base, __sz size)
 	 * way too late for the interrupt controller which is required for
 	 * bringing up secondary cores.
 	 *
-	 * Notice that we handle each page independently, as ukplat_page_map()
-	 * would return EEXIST even if part of a multi-pgae region is mapped,
-	 * causing the subsequent ukplat_page_set_attr() to fail.
+	 * Notice that updating the attributes of an already mapped region
+	 * can potentialy fail, as ukplat_page_map() will also return EEXIST
+	 * for a partially mapped region too. Addressing this problem by
+	 * remapping a region page-by-page is too slow for devices with large
+	 * regions like PCI.
 	 */
-	for (__sz i = 0; i < pages; i++) {
-		va = vaddr + i * PAGE_SIZE;
-		pa = paddr + i * PAGE_SIZE;
-		rc = ukplat_page_map(pt, va, pa, 1, attr, 0);
-		if (!rc)
-			continue;
+	rc = ukplat_page_map(pt, vaddr, paddr, pages, attr, 0);
+	if (unlikely(rc && rc != -EEXIST)) {
+		uk_pr_err("Could not map MMIO Region at 0x%lx - 0x%lx (%d)\n",
+			  base, base + size, rc);
+		return rc;
+	}
 
-		if (rc == -EEXIST)
-			rc = ukplat_page_set_attr(pt, va, 1, attr, 0);
+	if (rc == -EEXIST) {
```

Comment:
Will this still be needed after the ukboot rework? Who maps this zone so early that results in an EEXIST?

Reply:
At the current state of things probably nothing, but that should be irrelevant. As the function's scope of similar to that of `ioremap()` it should take that case into account. Additionally it may be called by some driver in the future at a later stage of execution.